File: aricode/sparring/challenges/13_minimax/minimax.go
// Go - Challenge 13: Minimax Game AI
// Nim(15): full tree search, return optimal move (3)
package main

import "os"

func minimax(stones int, maximizing bool) int {
	if stones == 0 {
		if maximizing { return -1 }
		return 1
	}
	if maximizing {
		best := -100
		for take := 1; take <= 3 && take <= stones; take++ {
			if s := minimax(stones-take, false); s > best { best = s }
		}
		return best
	}
	best := 100
	for take := 1; take <= 3 && take <= stones; take++ {
		if s := minimax(stones-take, true); s < best { best = s }
	}
	return best
}

func main() {
	bestMove, bestScore := 1, -100
	for take := 1; take <= 3; take++ {
		if s := minimax(15-take, false); s > bestScore {
			bestScore = s
			bestMove = take
		}
	}
	os.Exit(bestMove)
}

File: aricode/sparring/challenges/13_minimax/minimax.py
#!/usr/bin/env python3
# Python - Challenge 13: Minimax Game AI
# Nim(15): full tree search, return optimal move (3)
import sys
sys.setrecursionlimit(1000000)

def minimax(stones, maximizing):
    if stones == 0:
        return -1 if maximizing else 1
    if maximizing:
        best = -100
        for take in range(1, min(4, stones+1)):
            best = max(best, minimax(stones-take, False))
        return best
    else:
        best = 100
        for take in range(1, min(4, stones+1)):
            best = min(best, minimax(stones-take, True))
        return best

best_move, best_score = 1, -100
for take in range(1, 4):
    score = minimax(15 - take, False)
    if score > best_score:
        best_score = score
        best_move = take

sys.exit(best_move)
